src/macaron/malware_analyzer/pypi_heuristics/sourcecode/suspicious_setup.py

```diff
@@ -32,86 +33,86 @@ def _get_setup_source_code(self, api_client: PyPIRegistry) -> str | None:
+        if sourcecode_url is None:
+            return None
+
+        # Create a temporary directory to store the downloaded source.
         with tempfile.TemporaryDirectory() as temp_dir:
-            sourcecode_url: str | None = api_client.get_sourcecode_url()
-            if sourcecode_url is None:
-                return None
             try:
                 response = requests.get(sourcecode_url, stream=True, timeout=40)
                 response.raise_for_status()
             except requests.exceptions.HTTPError as http_err:
                 logger.debug("HTTP error occurred: %s", http_err)
                 return None
 
-            with open(os.path.join(temp_dir, "tmp"), "wb") as f:
-                f.write(response.content)
+            if response.status_code != 200:
+                return None
 
-            files = os.listdir(temp_dir)
+            source_file = os.path.join(temp_dir, "tmp")
+            with open(source_file, "wb") as file:
+                try:
+                    for chunk in response.iter_content():
+                        file.write(chunk)
+                except RequestException as error:
+                    # Something went wrong with the request, abort.
+                    logger.debug("Error while streaming source file: %s", error)
+                    response.close()
+                    return None
+
+            target_file = "setup.py"
             try:
-                with tarfile.open(os.path.join(temp_dir, files[0]), "r:gz") as tar:
-                    tar.extractall(temp_dir)  # nosec B202
+                with tarfile.open(source_file, "r:gz") as tar:
+                    tar.extract(target_file)
             except tarfile.ReadError as exception:
                 logger.debug("Error reading tar file: %s", exception)
 
                 try:
-                    with zipfile.ZipFile(os.path.join(temp_dir, files[0]), "r") as zip_ref:
-                        zip_ref.extractall(temp_dir)  # nosec B202
+                    with zipfile.ZipFile(source_file, "r") as zip_ref:
+                        zip_ref.extract(target_file)
                 except zipfile.BadZipFile as bad_zip_exception:
                     logger.debug("Error reading zip file: %s", bad_zip_exception)
                 except zipfile.LargeZipFile as large_zip_exception:
                     logger.debug("Zip file too large to read: %s", large_zip_exception)
 
-            for root, _, files in os.walk(temp_dir):
-                for file in files:
-                    if file == "setup.py":
-                        file_path = os.path.join(root, file)
-                        with open(file_path, encoding="utf-8") as py_file:
-                            return py_file.read()
-            return None
+            final_path = os.path.join(temp_dir, target_file)
+            if not os.path.exists(final_path):
+                logger.debug("Failed to download source file.")
```

Comment:
nit: I think it's better if we say we failed to find/extract `setup.py` from the source tarball as it had been downloaded earlier.